blog/content/post/2023-06-07-twpm-spi-fix.md

```diff
@@ -0,0 +1,855 @@
+	       <&dma1 4 1 0x20480>;
+	dma-names = "tx", "rx";
+};
+
+&spi3 {
+	status = "disabled";
+};
+```
+
+If we run this code we will see SPI errors running down the console:
+
+```shell
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+```
+
+This is the first problem with Zephyr's SPI driver - each transfer has a
+one-second timeout. While this may be desirable behavior for SPI master (it
+could be used for error recovery, for example, to power cycle the slave if it
+doesn't respond), it breaks SPI slave. The slave must be ready to give a
+response when the transfer commences - appropriate data must already be loaded
+in FIFO. Here we get stuck in an endless loop, queuing the transfer, aborting
+it, and queuing it again.
+
+The problem can be worked around by patching the `wait_dma_rx_tx_done` function
+in `spi_ll_stm32.c`. The original function looks like this:
+
+```c
+static int wait_dma_rx_tx_done(const struct device *dev)
+{
+	struct spi_stm32_data *data = dev->data;
+	int res = -1;
+
+	while (1) {
+		res = k_sem_take(&data->status_sem, K_MSEC(1000));
+		if (res != 0) {
+			return res;
+		}
+...
+```
+
+Just replace `K_MSEC(1000)` with `K_FOREVER`.
+
+Now running `spitest` at 100 KHz yields the following result:
+
+![STM32 SPI slave at 100 KHz](/img/stm32-spislave-100khz.png)
+
+The transfer works properly at 100 KHz. At 10 MHz the transfer sometimes works,
+sometimes does not:
+
+![STM32 SPI slave at 10 MHz](/img/stm32-spislave-10mhz.png)
+
+At 24 MHz transfer is completely corrupted. We have been looking for a solution
+in Zephyr Issues and Pull Requests but found nothing useful.
+
+Looking at Zephyr's
+SPI driver code, we discovered that every call to `spi_write` causes many things
+to happen. Among others, the SPI controller is reconfigured
+[every single time](https://github.com/zephyrproject-rtos/zephyr/blob/39391b4a160a4e23a2b7f213f94cf04b2c250ad7/drivers/spi/spi_ll_stm32.c#L751). During this process, the SPI controller is disabled and re-enabled,
+which is quite suspicious.
+
+## Reading STM32 documentation
+
+I've been searching through STM32 documentation for information about high-speed
+SPI. The most helpful were the STM32L4 series programming manual and
+.
+Section 4.2 of AN5543 describes various aspects of handling high-speed
+communication, and section 4.2.2 describes what happens when SPI is disabled.
+
+
+
+The main problem here is that Zephyr (as well as STM32 HAL) re-configures SPI
+before each transaction, doing configure-enable-transmit-disable cycle on each
+SPI session. While this is ok for master, slave must respect timings imposed
+by master, so SPI disabling should be avoided if not needed.
+
+The problem becomes even more pronounced when we want to implement TPM protocol
+as we don't know size (and direction) of data payload. Each TPM frame starts
+with a 4 byte header which tells us what is the size of transfer and what is the
```

Comment:
Consider adding graphic to explain that. Maybe bytefield from https://kroki.io/ (or similar) could be used there.

Reply:
I added image of TPM frame from PTP spec